frontend/src/hooks/useFetchData.ts

```diff
@@ -0,0 +1,151 @@
+import { RefObject, useCallback, useMemo, useState } from "react";
+import {
+  AgGridSortDirection,
+  DashboardRecordColumnFilter,
+  DashboardRecordSort,
+  DashboardRequestsDocument,
+  DashboardSamplesQueryVariables,
+} from "../generated/graphql";
+import { parseUserSearchVal } from "../utils/parseSearchQueries";
+import {
+  IServerSideDatasource,
+  IServerSideGetRowsParams,
+} from "ag-grid-community";
+import { AgGridReact as AgGridReactType } from "ag-grid-react/lib/agGridReact";
+import { usePhiEnabled } from "../contexts/PhiEnabledContext";
+import { LazyQueryHookOptions, useLazyQuery } from "@apollo/client";
+import { CACHE_BLOCK_SIZE } from "../config";
+
+function useHookLazyGeneric(baseOptions?: LazyQueryHookOptions<any, any>) {
+  const options = { ...({} as const), ...baseOptions };
+  return useLazyQuery<any, any>(DashboardRequestsDocument, options);
+}
+
+interface UseFetchDataParams {
+  useRecordsLazyQuery: typeof useHookLazyGeneric;
+  queryName: string;
+  initialSortFieldName: string;
+  gridRef: RefObject<AgGridReactType>;
+  userSearchVal: string;
+  contexts?: DashboardSamplesQueryVariables["contexts"];
+  pollInterval?: number;
+}
+
+export function useFetchData({
+  useRecordsLazyQuery,
+  queryName,
+  initialSortFieldName,
+  gridRef,
+  userSearchVal,
+  contexts = [],
+  pollInterval = 0, // 0 means no polling
+}: UseFetchDataParams) {
+  // Manage our own loading state becase the lazy query's provided `loading` state
+  // does not toggle to `true` as `setServerSideDatasource` is running
+  const [isLoading, setIsLoading] = useState(false);
+  const { phiEnabled } = usePhiEnabled();
+
+  const defaultSort: DashboardRecordSort = useMemo(
+    () => ({
+      colId: initialSortFieldName,
+      sort: AgGridSortDirection.Desc,
+    }),
+    [initialSortFieldName]
+  );
+
+  const [
+    ,
+    { error, data, fetchMore, refetch, startPolling: poll, stopPolling },
+  ] = useRecordsLazyQuery({
+    variables: {
+      searchVals: [],
+      contexts,
+      sort: defaultSort,
+      limit: CACHE_BLOCK_SIZE,
+      offset: 0,
+    },
+    pollInterval,
+  });
+
+  // Start poll using the same interval throughout the page
+  function startPolling() {
+    poll(pollInterval);
+  }
+
+  const recordCount: number = data?.[queryName][0]?._total || 0;
+
+  const buildServerSideDatasource = useCallback(
+    (userSearchVal) => {
+      return {
+        getRows: async (params: IServerSideGetRowsParams) => {
+          setIsLoading(true);
+
+          const variables = {
+            searchVals: parseUserSearchVal(userSearchVal),
+            contexts,
```

Comment:
to-do: clarify `contexts` (samples page has its own "context" indicating which filter option to use (i.e., wes vs. access), whereas `context` in other places indicates "requests", "cohorts", "patients" instead. 

suggesting a name change for clarity for one or both uses of "context"

Reply:
Agreed. Thoughts on `recordContexts` here?

Reply:
Done